internal/local/archive.go

```diff
@@ -0,0 +1,207 @@
+package local
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// set of supported archive file extensions
+const (
+	extTar   = ".tar"
+	extTarGz = ".tar.gz"
+	extTgz   = ".tgz"
+	extZip   = ".zip"
+)
+
+var (
+	supportedExts = []string{extZip, extTar, extTgz, extTarGz}
+)
+
+// ArchiveReader provides sequential access to the contents of the supported archive formats
+type ArchiveReader interface {
+	Next() (FileHeader, error)
+	Read(b []byte) (int, error)
+}
+
+// FileHeader holds the path and file information
+type FileHeader struct {
+	Path string
+	Info os.FileInfo
+}
+
+func newArchiveReader(path string, file *os.File) (ArchiveReader, error) {
+	ext := strings.ToLower(filepath.Ext(path))
+
+	if ext == extZip {
+		return newZipReader(file)
+	}
+
+	if ext == extTar {
+		return newTarReader(file), nil
+	}
+
+	if ext == extTgz || strings.HasSuffix(path, extTarGz) {
+		return newGzipReader(file)
+	}
+
+	if ext == "" {
+		return newDirReader(path)
+	}
```

Comment:
~[nit] this could just become a switch statement~ very disappointed we do this check `strings.HasSuffix(path, extTarGz)` which prevents us from doing a switch statement 😭

Reply:
agreed with you on that one